ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala

```diff
@@ -0,0 +1,216 @@
+            submission,
+            submission.getInputLogEntriesList.asScala
+              .map(eid => eid -> getLogEntry(state, eid))(breakOut),
+            submission.getInputDamlStateList.asScala
+              .map(key => key -> getDamlState(state, key))(breakOut)
+          )
+
+          // Combine the abstract log entry and the state updates into concrete updates to the store.
+          val allUpdates =
+            damlStateUpdates.map {
+              case (k, v) =>
+                NS_DAML_STATE.concat(KeyValueCommitting.packDamlStateKey(k)) ->
+                  KeyValueCommitting.packDamlStateValue(v)
+            } + (entryId.getEntryId -> KeyValueCommitting.packDamlLogEntry(logEntry))
+
+          // Atomically update the state reference, recording the commit
+          // and the updated state.
+          stateRef = state.copy(
+            log = state.log :+ (entryId -> submission),
+            store = state.store ++ allUpdates
+          )
+
+          // Wake up consumers.
+          dispatcher.signalNewHead(stateRef.log.size)
+        }
+    }
+  }
+  private val commitActorRef =
+    system.actorOf(Props(new CommitActor), s"commit-actor-${ledgerId.underlyingString}")
+
+  // Dispatcher to subscribe to 'Update' events derived from the state.
+  private val beginning: Int = 0
+  private val dispatcher: Dispatcher[Int, Update] = Dispatcher(
+    steppingMode = OneAfterAnother(
+      (idx: Int, _) => idx + 1,
+      (idx: Int) => Future.successful(getUpdate(idx, stateRef))
+    ),
+    zeroIndex = beginning,
+    headAtInitialization = beginning
+  )
+
+  private def getUpdate(idx: Int, state: State): Update = {
+    if (idx < 0 || idx >= state.log.size)
+      sys.error(s"getUpdate: $idx out of bounds (${state.log.size})")
+
+    // Resolve the "height" to log entry.
+    val entryId = state.log(idx)._1
+
+    state.store
+      .get(entryId.getEntryId)
+      .map { blob =>
+        KeyValueConsumption.logEntryToUpdate(entryId, KeyValueConsumption.unpackDamlLogEntry(blob))
+      }
+      .getOrElse(
+        sys.error(s"getUpdate: $entryId not found from store!")
+      )
+  }
+
+  override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] = {
+    dispatcher
+      .startingAt(beginAfter.map(_.components.head.toInt + 1).getOrElse(beginning))
+      .map {
+        case (idx, update) =>
+          Offset(Array(idx.toLong)) -> update
+      }
+  }
+
+  private def getLogEntry(state: State, entryId: DamlLogEntryId): DamlLogEntry =
+    DamlLogEntry.parseFrom(state.store(entryId.getEntryId))
+
+  private def getDamlState(state: State, key: DamlStateKey): Option[DamlStateValue] =
+    state.store
+      .get(NS_DAML_STATE.concat(KeyValueCommitting.packDamlStateKey(key)))
+      .map(DamlStateValue.parseFrom)
+
+  override def submitTransaction(
+      submitterInfo: SubmitterInfo,
+      transactionMeta: TransactionMeta,
+      transaction: SubmittedTransaction): Unit = {
+
```

Comment:
extend comments on what's happening -- I'm assuming this should be the example that explains how to implement the participant-state interfaces using the kvutils.